2_NeuralNetworkClassification/Exercises2.py

```diff
@@ -0,0 +1,233 @@
+import torch
+import matplotlib.pyplot as plt
+from sklearn.model_selection import train_test_split
+from torch import nn
+import sys
+import os
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
+from helper_functions import plot_predictions, plot_decision_boundary
+from sklearn.datasets import make_moons
+
+# Create device agnostic code
+device = "cuda" if torch.cuda.is_available() else "cpu"
+print(device)
+print("\n")
+
+def accuracy_fn(y_true, y_pred):
+    correct = torch.eq(y_true, y_pred).sum().item() # torch.eq() calculates where two tensors are equal
+    acc = (correct / len(y_pred)) * 100 
+    return acc
+
+# Make a binary classification dataset with Scikit-Learn's make_moons() function.
+# For consistency, the dataset should have 1000 samples and a random_state=42.
+# Turn the data into PyTorch tensors. Split the data into training and test sets 
+# using train_test_split with 80% training and 20% testing.
+
+# Make 1000 samples 
+n_samples = 1000
+# Create circles
+X, y = make_moons(n_samples,
+                    noise=0.03,
+                    random_state=42)
+
+# Visualize with a plot
+plt.figure(figsize=(10, 7))
+plt.scatter(x=X[:, 0], 
+            y=X[:, 1], 
+            c=y, 
+            cmap=plt.cm.RdYlBu)
+# Save the plot as an image 
+plt.savefig('2_NeuralNetworkClassification/Ex_moons_dataset.png')
+print("Plot saved as 'Ex_moons_dataset.png'")
+plt.clf()
+
+# Turn data into tensors
+# Otherwise this causes issues with computations later on
+import torch
+X = torch.from_numpy(X).type(torch.float)
+y = torch.from_numpy(y).type(torch.float)
+
+# View the first five samples
+print(X[:5], y[:5], "\n")
+
+# Split data into train and test sets
+from sklearn.model_selection import train_test_split
+
+X_train, X_test, y_train, y_test = train_test_split(X, 
+                                                    y, 
+                                                    test_size=0.2, # 20% test, 80% train
+                                                    random_state=42) # make the random split reproducible
+
+print(len(X_train), len(X_test), len(y_train), len(y_test))
+print("\n")
+
+
+# Build a model by subclassing nn.Module that incorporates non-linear activation 
+# functions and is capable of fitting the data you created in 1.
+
+# Build model with non-linear activation function
+from torch import nn
+class MoonModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.layer_1 = nn.Linear(in_features=2, out_features=10)
+        self.layer_2 = nn.Linear(in_features=10, out_features=10)
+        self.layer_3 = nn.Linear(in_features=10, out_features=1)
+        self.relu = nn.ReLU()
```

Comment:
does it make sense to have 3 consecutive linear layers with no activation function in between?